### ISSUE TITLE
Support "Go: Autocomplete Unimported Packages" feature when using Go modules

### ISSUE THREAD

**Issue by ramya-rao-a** [microsoft/vscode-go#2484]
Enabling the setting `go.autocompleteUnimportedPackages` used to provide completions for unimported packages. When such a completion item was selected by the user, not only was the current word completed, but an import statement is added in the file for that package.

This was implemented by going over all the packages under the GOPATH and keeping a cache of this result.

With Go modules, it is not clear how we can provide this feature.

@stamblerre Is this something that `gopls` can do as part of the completion provider?

**Comment by wxdao** [comment]
Maybe at least make it to work with packages under the current module and those in vendor folder? They are presented in source code files and the good old going-over approach should work.

**Comment by stamblerre** [comment]
@ramya-rao-a, we are definitely planning on supporting this as part of `gopls`. I filed https://github.com/golang/go/issues/31906 to track it. It's not as high priority as some other features, but it is definitely on our list.

**Comment by ramya-rao-a** [comment]
Thanks @stamblerre 
Once this feature makes it into `gopls` (assuming that it will work both for non modules set up as well), please log an issue or create a PR here to remove the similar feature here to avoid duplicate suggestions

Closing this issue in favor of the upstream issue.

**Comment by ramya-rao-a** [comment]
Oh closed too soon!

> Maybe at least make it to work with packages under the current module and those in vendor folder? They are presented in source code files and the good old going-over approach should work.

That is an interesting take, how do you suggest we get the list of such packages? Would `go list` provide this information?

**Comment by fintanchen** [comment]
I think there is a way that caching standard library's packages information to some place, because everybody would use it everyday and itself isn't change too often. And provide the list of dependencies's by parsing `go.mod`.

**Comment by paivagustavo** [comment]
Doesn't `go list -m all` suffices for this case? It lists dependencies, including indirect ones.

**Comment by paivagustavo** [comment]
With `-m` flag we only get the list of modules. If we drop it out, like `go list all`, we get the list of all packages, that is what we may need for this.

On modules it includes all dependencies' packages and standard library's packages as well.

**Comment by ramya-rao-a** [comment]
Thanks @fintanchen and @paivagustavo 
@stamblerre Is that what you would suggest to get list of packages that can be considered as valid imports?

**Comment by ianthehat** [comment]
There is no fast way to do what you are trying to do here with existing tools.
If you want to see the kind of logic you need, you can look at the things that goimports does to try to find new imports, which is essentially the task you are trying to do here as well.
I don't think there is anything you could reasonable do here that it would not be easier and better to just add directly to gopls instead.

**Comment by ramya-rao-a** [comment]
Thanks @ianthehat 

If anyone wants to help with the issue, please look at https://github.com/golang/go/issues/31906 to help contribute a way to get this feature from `gopls`

I'll keep this issue open just in case if anyone wants to give the `go list` option a try.

**Comment by colinlabs** [comment]
[saibing/tools](https://github.com/saibing/tools) support "Autocomplete Unimported Packages", can this help with the issue?

**Comment by chengyayu** [comment]
**versions**

- vscode : 1.39.2
- vscode-go: 0.11.7

**my settings**

```json
 // go settings
    "go.buildOnSave": "off",
    "go.lintTool":"golangci-lint",
    "go.lintFlags": [
        "--fast"
    ],
    "go.useLanguageServer": true,
    "go.languageServerExperimentalFeatures": {
        "documentLink": false,
        "format": true,
        "autoComplete": true
    },
    "[go]": {
        "editor.snippetSuggestions": "none",
        "editor.formatOnSave": true,
        "editor.codeActionsOnSave": {
            "source.organizeImports": true
        }
    },
    "gopls": {
        "usePlaceholders": true,
        "completeUnimported": true,
    },
```

when i type this codes go internal package can be auto import, but third party package in my go.mod can not be import.

**gopls output error**

```
[Info  - 5:23:18 PM] 2019/10/22 17:23:18 no completions found
	Failure = cannot resolve grpc
```

**Comment by stamblerre** [comment]
golang/go#31906 has been closed upstream, and this will be supported by default in `gopls` with `gopls/v0.3.0`, so I think that this issue can be resolved.

**Comment by ramya-rao-a** [comment]
That's great news @stamblerre 

Is there a way one can disable this feature?
When using this feature from the extension, one has to explicitly opt in using the `go.autocompleteUnimportedPackages` setting.
The reason behind doing so was that there was early feedback on this feature being noisy for some.

**Comment by stamblerre** [comment]
It can be disabled by setting the `"completeUnimported"` value to `false`, but this feature has been refined a lot in `gopls`, so I expect that it will work a lot better for people than `gocode` did. `gopls` has careful ranking and relies on an exact package name match, so hopefully few people will want to turn it off.

**Comment by ramya-rao-a** [comment]
Thanks @stamblerre 

Closing this issue as the feature will be available in the next version `gopls/v0.3.0` of the language server. Folks who want to try the feature out before the release can do so by downloading gopls by running
```
GO111MODULE=on 
go get golang.org/x/tools/gopls@master golang.org/x/tools@master
```
